lib/domain-page/ensService.ts

```diff
@@ -0,0 +1,175 @@
+import {
+  batch,
+  getAvailable,
+  getExpiry,
+  getName,
+  getOwner,
+  getRecords,
+  getResolver,
+} from "@ensdomains/ensjs/public";
+import { getSubgraphRecords } from "@ensdomains/ensjs/subgraph";
+import { GraphQLClient } from "graphql-request";
+import { normalize } from "viem/ens";
+import {
+  getCoinNameByType,
+  getSupportedCoins,
+  transformTextRecords,
+  updateAvatarInTexts,
+  validateDomain,
+} from "./utils";
+import {
+  DomainData,
+  QueryDomain,
+  QueryResponse,
+  SubgraphEnsData,
+} from "./interfaces";
+import { query } from "./queries";
+import { ContractFunctionExecutionError, parseAbiItem } from "viem";
+import toast from "react-hot-toast";
+
+// Ensure API key is available
+const ensSubgraphApiKey = process.env.NEXT_PUBLIC_ENS_SUBGRAPH_KEY;
+if (!ensSubgraphApiKey) {
+  throw new Error("ENS subgraph API key not found");
+}
+
+// Fetch ENS data for a given domain
+export const fetchDomainData = async (
+  domain: string
+): Promise<DomainData | null> => {
+  // Check if the domain resolver is compatible with metadata, if nor it will return an error
+  try {
+    const data = await fetchDomainDataThroughResolver(domain);
+    const domainData = formatDomainDataThroughResolver(data);
+    return domainData;
+    // Case where it's not compatilble
+  } catch (error) {
+    if (error instanceof ContractFunctionExecutionError) {
+      const data = await getENSDomainDataThroughSubgraph(domain);
+      if (!data) return null;
+      const domainData = await formatENSDomainDataThroughSubgraph(data);
+      return domainData;
+    } else {
+      toast.error("An Error occurred while loading the data");
+    }
+    return null;
+  }
+};
+
+// Fetch ENS data for a given domain through subgraph
+export const getENSDomainDataThroughSubgraph = async (
+  domain: string
+): Promise<SubgraphEnsData | null> => {
+  validateDomain(domain);
+
+  if (await getAvailable(client, { name: domain })) return null;
+
+  const textRecordsKeys = await getSubgraphRecords(client, { name: domain });
+  const availableTextRecords = textRecordsKeys?.texts.length
+    ? textRecordsKeys?.texts
+    : defaultTextRecords;
+
+  const [newAvatar, batchResults] = await Promise.all([
+    publicClient.getEnsAvatar({ name: normalize(domain) }),
+    batch(
+      client,
+      getRecords.batch({
+        name: domain,
+        texts: availableTextRecords,
+        coins: getSupportedCoins(),
+        contentHash: true,
+      }),
+      getOwner.batch({ name: domain }),
+      getExpiry.batch({ name: domain }),
+      getResolver.batch({ name: domain })
+    ),
+  ]);
+
+  const [textRecords, owner, expiry] = batchResults;
+
+  const data: SubgraphEnsData = {
+    newAvatar,
+    ...textRecords,
+    ...owner,
+    owner: owner?.owner ?? "0x",
+    ...expiry,
+  };
+
+  return data;
+};
+
+const fetchDomainDataThroughResolver = async (
+  name: string
+): Promise<QueryDomain> => {
+  const resolverAdd = await getResolver(publicClient, { name });
+
```

Comment:
@FrancoAguzzi At this point, the function:

 ```
const metadataUrl = await publicClient.readContract({
    address: resolverAdd!,
    abi: [parseAbiItem("function metadata() returns (string)")],
    functionName: "metadata",
  });
```
already returns an error if the resolver doesn't support metadata calls. When dealing with not a resolver statuses, we likely won't handle that check within this function, keeping each function focused and independent.

Reply:
@eduramme very nice! I like it!

But the exclamation is yet a question for me: can you write a comment above it saying that if resolverAdd is null, the function will return an error which will be handled somehow?

Actually I just noticed that this function would not throw an explainative error for whoever called it because it is not inside a try catch block, can you add it? pls :) 🙏🏼 Then the comment mentioned can say that whenever resolverAdd is null, an error will be thrown by "readContract" call

I think the function will look shiny then + will share the error with the parent call

WDYT?